crates/holochain/src/conductor/conductor/tests.rs

```diff
@@ -1224,13 +1215,79 @@ async fn test_deferred_provisioning() {
 
     //- Can't make zome calls, error returned is CellDisabled
     //  (which isn't ideal, but gets the message across well enough)
-    let result: Result<ActionHash, _> = conductor
-        .call_fallible(&cell.zome("zome"), "create", ())
-        .await;
+    let result: Result<String, _> = conductor.call_fallible(&cell.zome("foo"), "foo", ()).await;
     assert_matches!(
         result,
         Err(ConductorApiError::ConductorError(
             ConductorError::CellDisabled(_)
         ))
     );
+
+    conductor.shutdown().await;
+    conductor.startup().await;
+
+    //- Status is still AwaitingMemproofs after a restart
+    let app_info = conductor.get_app_info(&app_id).await.unwrap().unwrap();
+    assert_eq!(app_info.status, AppInfoStatus::AwaitingMemproofs);
+
+    //- Status is still AwaitingMemproofs after enabling but before memproofs
+    conductor.enable_app(app_id.clone()).await.unwrap();
+    let app_info = conductor.get_app_info(&app_id).await.unwrap().unwrap();
+    assert_eq!(app_info.status, AppInfoStatus::AwaitingMemproofs);
+
+    //- Rotate app agent key a few times just for the heck of it
+    // TODO: not yet implemented
+
+    // let agent1 = conductor.rotate_app_agent_key(&app_id).await.unwrap();
+    // let agent2 = conductor.rotate_app_agent_key(&app_id).await.unwrap();
+    // let agent3 = conductor.rotate_app_agent_key(&app_id).await.unwrap();
+    // assert_ne!(agent_key, agent1);
+    // assert_ne!(agent1, agent2);
+    // assert_ne!(agent2, agent3);
+
+    //- Now provide the memproofs
+    conductor
+        .clone()
+        .provide_memproofs(&app_id, MemproofMap::new())
+        .await
+        .unwrap();
+
+    conductor.enable_app(app_id.clone()).await.unwrap();
+
+    //- Status is now Running and there is 1 cell assignment
+    let app_info = conductor.get_app_info(&app_id).await.unwrap().unwrap();
+    assert_eq!(app_info.status, AppInfoStatus::Running);
+
+    //- And now we can make a zome call successfully
+    let _: String = conductor.call(&cell.zome("foo"), "foo", ()).await;
+}
+
+/// Can uninstall an app with deferred memproofs before providing memproofs
+#[tokio::test(flavor = "multi_thread")]
+async fn test_deferred_provisioning_uninstall() {
+    holochain_trace::test_run();
+    let (dna, _, _) = SweetDnaFile::unique_from_test_wasms(vec![TestWasm::Foo]).await;
+    let conductor = SweetConductor::from_standard_config().await;
+    let app_id = "app-id".to_string();
+    let agent_key = SweetAgents::one(conductor.keystore()).await;
+    let role_name = "role".to_string();
+    let bundle = app_bundle_from_dnas([&(role_name.clone(), dna)]).await;
+
+    //- Install with deferred memproofs
+    conductor
+        .clone()
+        .install_app_bundle(InstallAppPayload {
+            source: AppBundleSource::Bundle(bundle),
+            agent_key: agent_key.clone(),
+            installed_app_id: Some(app_id.clone()),
+            membrane_proofs: MemproofProvisioning::Deferred,
```

Comment:
Hmm...looking at this `InstallAppPayload` I think we're still missing a part. How should Launcher know what MemproofProvisioning variant to choose when installing an app? I thought we would add a field in the happ manifest for this, in which case the `membrane_proofs` field could also potentially just be an `Option<MembraneProofs>` rather than an enum? Or just remain the same, i.e. no Option and they are ignored by the conductor if the manifest says deferred.

Reply:
We could also add a third enum variant `MemproofProvisioning::Manifest` which would let it up to the conductor to pick the right strategy based on the app manifest but still leave the freedom to launcher to override if necessary.

Reply:
The same could be achieved by keeping the existing enum but converting the `membrane_proofs` field to an `Option<MembraneProvisioning>`. Providing something will then always be considered an override of the manifest.

Reply:
You're right, I missed this detail. I added it in in a followup PR (even though it reverts a bunch of changes from the previous PR) https://github.com/holochain/holochain/pull/3944